### PR TITLE
[SPARK-6107][CORE] Display inprogress application information for event log history for standalone mode

### PR DESCRIPTION
**Pull request by liyezhang556520**
when application is finished running abnormally (Ctrl + c for example), the history event log file is still ends with `.inprogress` suffix. And the application state can not be showed on webUI, User can only see "_Application history not foud xxxx, Application xxx is still in progress_". 

For application that not finished normally, the history will show:
image

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -736,30 +736,31 @@ private[spark] class Master(
...
-      if (fs.exists(new Path(eventLogFile + EventLoggingListener.IN_PROGRESS))) {
+      
+      val eventLogFilePrefix = EventLoggingListener.getLogPath(eventLogDir, app.id)        
+      val fs = Utils.getHadoopFileSystem(eventLogDir, hadoopConf)
+      val eventLogFileSuffix = if (fs.exists(new Path(eventLogFilePrefix + 
+          EventLoggingListener.IN_PROGRESS))) {
         // Event logging is enabled for this application, but the application is still in progress
-        val title = s"Application history not found (${app.id})"
-        var msg = s"Application $appName is still in progress."
+        var msg = s"Application $appName is still in progress, it may be terminated accidently."
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: "accidently" is misspelled. I think this line may be too long. Also, it's probably unusual that the app terminated right? this file will always show up while the app is running?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 2]
```diff
@@ -736,30 +736,31 @@ private[spark] class Master(
...
         logWarning(msg)
-        msg = URLEncoder.encode(msg, "UTF-8")
-        app.desc.appUiUrl = notFoundBasePath + s"?msg=$msg&title=$title"
-        return false
-      }
+        EventLoggingListener.IN_PROGRESS
+      } else ""
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: This seems harder to understand than just perhaps recording the value of `if (fs.exists(new Path(eventLogFilePrefix + EventLoggingListener.IN_PROGRESS)))` in a `val`, using it above in the `if`, and using it here to construct the file name.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Hi @srowen , when app is running, it will recording the history at the same time. And the history log file name is with the format "app-xxxxx.inprogress" and after the app finished running, the suffix (.inprogress) is removed. And if app is stopped abnormally, the suffix remains.

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Yeah I understand that, but the message tends to suggest to people that there is a problem, when 99% of the time the app is still merely running, not failed. I might soften the text to "... still in progress. If the application is not running, then this indicates it may have terminated abnormally."

Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Ok, I'll update the message, thank you for your idea

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 2 above)
Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Hi @srowen , I don't quite understand your idea, Can you explain it in detail? I just intend not to introduce `var`
>using it above in the if, and using it here to construct the file name

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Like:

`val inProgressExists = fs.exists(new Path(eventLogFilePrefix + EventLoggingListener.IN_PROGRESS))`

Then use that in the `if` statement, which need not also evaluate to a value.
Then use the same flag to construct the file name.
IMHO this is more straightforward than piggy-backing a sometimes-blank return value from the conditional.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: @srowen , I have a question, is it normal that the app is still running after it is removed by master ? Should not it  be always be stopped when master receive `removeApplication` message?

Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I don't think that's normal. Hm, is that the context here? I thought this was just about always seeing the latest log messages even from the in-progress log file.

Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Well I don't think the in-progress log file will be appended any more log messages when the program comes here.

**Comment by liyezhang556520** [comment]
Hi @srowen , I updated the code, can you have a look. Please correct me if I still misunderstand your idea.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 3]
```diff
@@ -736,30 +736,34 @@ private[spark] class Master(
...
+      val eventLogFilePrefix = EventLoggingListener.getLogPath(eventLogDir, app.id)        
+      val fs = Utils.getHadoopFileSystem(eventLogDir, hadoopConf)
+      val inProgressExists = fs.exists(new Path(eventLogFilePrefix + 
+          EventLoggingListener.IN_PROGRESS))
+      
+      if (inProgressExists) {
         // Event logging is enabled for this application, but the application is still in progress
-        val title = s"Application history not found (${app.id})"
-        var msg = s"Application $appName is still in progress."
+        var msg = s"Application $appName is still in progress, it may be terminated accidently."
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: This could be a `val` but could also just be an arg to `logWarning`.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 4]
```diff
@@ -736,30 +736,34 @@ private[spark] class Master(
...
         logWarning(msg)
-        msg = URLEncoder.encode(msg, "UTF-8")
-        app.desc.appUiUrl = notFoundBasePath + s"?msg=$msg&title=$title"
-        return false
       }
+      
+      val eventLogFile = eventLogFilePrefix + {
+        if(inProgressExists) EventLoggingListener.IN_PROGRESS
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Oops, meant to comment here. I don't believe this works since you don't have an else clause. If the condition it false it appends something random like `()`. I'd keep it straightforward.

```
val eventLogFile  = if (inProgressExists) {
  ...
} else {
  ...
}

**Comment by srowen** [comment]
CC @viirya

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 4 above)
Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Yes, you are right, I discovered that, i'll update it soon, thanks.

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 5]
```diff
@@ -736,30 +736,34 @@ private[spark] class Master(
...
+        logWarning(s"Application $appName is still in progress, it may be terminated accidently.")
       }
-
+      
+      val eventLogFile = if (inProgressExists) {
+        eventLogFilePrefix + EventLoggingListener.IN_PROGRESS
+      } else {
+        eventLogFilePrefix
+      }
+      val status = if (inProgressExists) " (inprogress)" else " (completed)"
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: `status` is determined based on same condition as `eventLogFile`. Could be combined together in a if elese block.

Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Good idea, thans @viirya

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala (hunk 1 above)
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: I agreed with @srowen that the wording `...,it may be terminated accidently` seems suggesting there is a problem. Using softer wording would be better.

**Comment by liyezhang556520** [comment]
Jenkins, retest this please.

**Comment by liyezhang556520** [comment]
Hi @srowen , @viirya , any further comments?

**Comment by viirya** [comment]
Looks like there are conflicts for merging, merge the updates first?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 6]
```diff
@@ -736,30 +736,34 @@ private[spark] class Master(
...
-        logWarning(msg)
-        msg = URLEncoder.encode(msg, "UTF-8")
-        app.desc.appUiUrl = notFoundBasePath + s"?msg=$msg&title=$title"
-        return false
+        logWarning(s"Application $appName is still in progress, it may be terminated abnormally.")
       }
-
+      
+      val (eventLogFile, status) = if (inProgressExists) {
+        (eventLogFilePrefix + EventLoggingListener.IN_PROGRESS, " (inprogress)")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: One last comment -- I see that "inprogress" is used in the UI. It should be "in progress".

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 7]
```diff
@@ -774,7 +778,7 @@ private[spark] class Master(
       case fnf: FileNotFoundException =>
         // Event logging is enabled for this application, but no event logs are found
         val title = s"Application history not found (${app.id})"
-        var msg = s"No event logs found for application $appName in ${app.desc.eventLogDir}."
+        var msg = s"No event logs found for application $appName in ${app.desc.eventLogDir.get}."
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: the `.get` is a minor unrelated change, right?

I think this looks good in that it changes a situation where Spark does nothing and just shows an error, into one where it shows something, which may be needed for debugging.

Comment (liyezhang556520) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala]: Yes, it's an unrelated change.
